From 8abf51c8751eca0e2652dbd1cd5e9d404358da2a Mon Sep 17 00:00:00 2001
From: Frank Li <Frank.Li@nxp.com>
Date: Wed, 28 Jun 2017 11:36:57 -0500
Subject: [PATCH 2023/5242] MLK-15300-1 perf: fixed crash when irq have not
 assigned in dts file

commit  bc15f4960bb43a871fbd158cf1603bf284b75736 from
https://source.codeaurora.org/external/imx/linux-imx.git

[ 28.061044] perf[2494]: PC Alignment exception: pc=0000000072656d69 sp=ffff800032f2bd30
[ 28.069061] Internal error: Oops - SP/PC alignment exception: 8a000000 1 PREEMPT SMP
[ 28.077066] Modules linked in:
[ 28.080128] CPU: 2 PID: 2494 Comm: perf Not tainted 4.9.11-02540-g3ebe22c #52
[ 28.087263] Hardware name: Freescale i.MX8QXP LPDDR4 ARM2 (DT)
[ 28.093093] task: ffff80002e097080 task.stack: ffff800032f28000
[ 28.099011] PC is at 0x72656d69
[ 28.102163] LR is at perf_try_init_event+0x98/0xb0
[ 28.106950] pc : [<0000000072656d69>] lr : [<ffff0000081594e0>] pstate: 60000145

Signed-off-by: Frank Li <Frank.Li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/perf/ddr-perf.c |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/perf/ddr-perf.c b/drivers/perf/ddr-perf.c
index 2ec4ce9..cbebdf6 100644
--- a/drivers/perf/ddr-perf.c
+++ b/drivers/perf/ddr-perf.c
@@ -449,11 +449,13 @@ static int ddr_perf_probe(struct platform_device *pdev)
 					pmu);
 	if (ret < 0) {
 		pr_err("Request irq failed: %d", ret);
-		goto ddr_perf_err;
+		goto ddr_perf_irq_err;
 	}
 
 	return 0;
 
+ddr_perf_irq_err:
+	perf_pmu_unregister(&(pmu->pmu));
 ddr_perf_err:
 	pr_warn("i.MX8 DDR Perf PMU failed (%d), disabled\n", ret);
 	kfree(pmu);
-- 
1.7.9.5

